Add various "Expand ..." / "Collapse ..." context menu actions to the top-down view

dpallotti opened the pull request:
#### Move topDownView from OrbitMainWindow to new TopDownWidget
#### Add "Expand all", "Collapse all" context menu actions to TopDownWidget
Bug: http://b/162704044
#### Allow multiple selections in TopDownWidget
...and hence multiple nodes to be "expanded all" or "collapsed all" at once.
#### Use Qt 5 signal-slot syntax in TopDownWidget
#### Use header()->resizeSections instead of a loop on resizeColumnToContents
This doesn't change the functionality, only shortens the code.
#### Rename "Expand / Collapse all" to "Expand / Collapse recursively"
"Expand / Collapse all" seems to usually refer to the entire tree.
#### Add "Expand / Collapse all" (the real ones this time) to top-down view
#### Add "Collapse children recursively" to TopDownWidget

pierricgimmig commented on OrbitQt/topdownwidget.h:
```diff
@@ -0,0 +1,38 @@
+  explicit TopDownWidget(QWidget* parent = nullptr)
+      : QWidget{parent}, ui_{std::make_unique<Ui::TopDownWidget>()} {
+    ui_->setupUi(this);
+    connect(ui_->topDownTreeView,
```
Qt's new signal-slot syntax is preferable: https://wiki.qt.io/New_Signal_Slot_Syntax

dpallotti replied:
Done, thank you.

pierricgimmig commented:
Hi @dpallotti , thanks for this, code looks good but please see comment about signal-slot syntax. 

Concerning functionality, What would you think of the "collapse all" applying to child nodes only. I'd prefer it that way actually, because you can already close the entire node by clicking on the triangle toggle. It's useful to see the first level of functions under a specific node, with the current implementation, I'd have to "collapse all" then reopen the node.

pierricgimmig commented on OrbitQt/topdownwidget.cpp:
```diff
@@ -0,0 +1,104 @@
+  ui_->topDownTreeView->sortByColumn(TopDownViewItemModel::kInclusive,
+                                     Qt::DescendingOrder);
+  for (int column = 0; column < TopDownViewItemModel::kColumnCount; ++column) {
+    ui_->topDownTreeView->resizeColumnToContents(column);
```
This doesn't seem to work for the name column on my side, unless there is a maximum width specified somewhere?

dpallotti replied:
I think that this works, but not like we'd like because the resizing is done based on the visible elements, which when this code is called are only the thread names. I'm not sure what a good way of doing this is.

beckerhe replied:
You can use QHeaderView to achieve what you want (probably). Have a look in one of my branches: https://github.com/beckerhe/orbitprofiler/blob/9e362d1d4611c88ada629c57002cb71ca74b0227/OrbitQt/orbitmainwindow.cpp#L270

beckerhe replied:
And a doc link: https://doc.qt.io/qt-5/qheaderview.html#setSectionResizeMode-1

dpallotti replied:
I had actually tried that but if I use `Stretch` the behavior is weird. The section is not manually resizable anymore. You can resize it indirectly by resizing the other ones but as it's the first section (or rather, not the last) the overall behavior is annoying.

dpallotti commented:
> Concerning functionality, What would you think of the "collapse all" applying to child nodes only. I'd prefer it that way actually, because you can already close the entire node by clicking on the triangle toggle. It's useful to see the first level of functions under a specific node, with the current implementation, I'd have to "collapse all" then reopen the node.

I think we're looking at the problem slightly differently: I was thinking about the "Collapse all" as a way to close the entire subtree without having the problem of it completely re-expanding once the node is expanded. With your suggestion, to get that one has to perform two operations&mdash;collapse the children and then collapse the node.
In general I don't dislike the functionality, but I think it's not what one would expect as it's not symmetrical with "Expand all". So if we want it, we should rename it to "Collapse children recursively" or just add an additional entry with that name.

Speaking of naming, maybe these actions should indeed be called "Expand recursively" and "Collapse recursively". It seems that the common behavior of "Expand all" and "Collapse all" in other applications is to expand or collapse the entire tree, regardless of the selected item(s).

dpallotti commented:
> > Concerning functionality, What would you think of the "collapse all" applying to child nodes only. I'd prefer it that way actually, because you can already close the entire node by clicking on the triangle toggle. It's useful to see the first level of functions under a specific node, with the current implementation, I'd have to "collapse all" then reopen the node.
> 
> I think we're looking at the problem slightly differently: I was thinking about the "Collapse all" as a way to close the entire subtree without having the problem of it completely re-expanding once the node is expanded. With your suggestion, to get that one has to perform two operations—collapse the children and then collapse the node.
> In general I don't dislike the functionality, but I think it's not what one would expect as it's not symmetrical with "Expand all". So if we want it, we should rename it to "Collapse children recursively" or just add an additional entry with that name.
> 
> Speaking of naming, maybe these actions should indeed be called "Expand recursively" and "Collapse recursively". It seems that the common behavior of "Expand all" and "Collapse all" in other applications is to expand or collapse the entire tree, regardless of the selected item(s).

@pierricgimmig Based on this, I changed "Expand all" / "Collapse all" to "Expand recursively", "Collapse recursively", added "Collapse children recursively", added "Expand all" / Collapse all" to actually expand and collapse everything. Let me know what you think and if it's too many actions.

beckerhe commented on OrbitQt/topdownwidget.cpp:
```diff
@@ -0,0 +1,151 @@
+const std::string TopDownWidget::kActionExpandAll = "Expand all";
+const std::string TopDownWidget::kActionCollapseAll = "Collapse all";
+
+static void ExpandRecursively(QTreeView* tree_view, const QModelIndex& index) {
```
Maybe I missed it, but is there a reason not to use QTreeView::expandRecursively?

dpallotti replied:
You caught me off guard for a moment, as I had tried and I didn't remember, but the reason is that it should be there from Qt 5.13 but on my Linux machine I can only get Qt up to 5.12.5... So either I leave it like this or I'll need help.

dpallotti replied:
Also there's no `QTreeView::collapseRecursively` so might as well implement the expand version ourselves as well.

beckerhe commented on OrbitQt/topdownwidget.cpp:
```diff
@@ -0,0 +1,151 @@
+  }
+}
+
+static void ExpandAll(QTreeView* tree_view) {
```
As before. This seems to do the same as QTreeView::expandAll, isn't it?

beckerhe commented on OrbitQt/topdownwidget.cpp:
```diff
@@ -0,0 +1,151 @@
+  }
+}
+
+static void CollapseAll(QTreeView* tree_view) {
```
QTreeView::collapseAll?

beckerhe commented on OrbitQt/topdownwidget.h:
```diff
@@ -18,11 +18,20 @@ class TopDownWidget : public QWidget {
+  void onCustomContextMenuRequested(const QPoint& point);
+
  private:
+  static const std::string kActionExpandAll;
```
Edit: After seeing the rest of the code: Using QStringView or QStringLiteral would avoid some conversions.

dpallotti replied:
I went with `QString`, using the `QStringLiteral` macro to create the value. I'm not sure why you suggested `string_view`/`QStringView` here though. Aren't those supposed to be used as parameters to receive strings they don't own?

dpallotti commented on OrbitQt/topdownwidget.ui:
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ui version="4.0">
```
Exactly, the idea was to add a `QLineEdit` for filtering but then I delayed that as things got trickier than expected. So I'll leave it as the plan would be to get to it soon.

dpallotti commented on OrbitQt/topdownwidget.h:
```diff
@@ -18,11 +18,20 @@ class TopDownWidget : public QWidget {
       : QWidget{parent}, ui_{std::make_unique<Ui::TopDownWidget>()} {
     ui_->setupUi(this);
+    connect(ui_->topDownTreeView,
+            SIGNAL(customContextMenuRequested(const QPoint&)), this,
```
Yep, this is done in a subsequent commit as Pierric also pointed that out.

dpallotti commented on OrbitQt/topdownwidget.h:
```diff
@@ -0,0 +1,29 @@
+#ifndef ORBIT_QT_TOP_DOWN_WIDGET_H_
+#define ORBIT_QT_TOP_DOWN_WIDGET_H_
+
+#include <QSortFilterProxyModel>
```
Done.

dpallotti commented on OrbitQt/topdownwidget.cpp:
```diff
@@ -65,36 +67,38 @@ void TopDownWidget::onCustomContextMenuRequested(const QPoint& point) {
-  if (enable_expand_all) menu.addAction(kActionExpandAll.c_str());
-  if (enable_collapse_all) menu.addAction(kActionCollapseAll.c_str());
+  if (enable_expand_recursively)
+    menu.addAction(kActionExpandRecursively.c_str());
```
The constant is now a `QString`.

pierricgimmig commented:
> I think we're looking at the problem slightly differently: I was thinking about the "Collapse all" as a way to close the entire subtree without having the problem of it completely re-expanding once the node is expanded. With your suggestion, to get that one has to perform two operations—collapse the children and then collapse the node.
> In general I don't dislike the functionality, but I think it's not what one would expect as it's not symmetrical with "Expand all". So if we want it, we should rename it to "Collapse children recursively" or just add an additional entry with that name.
> 
> Speaking of naming, maybe these actions should indeed be called "Expand recursively" and "Collapse recursively". It seems that the common behavior of "Expand all" and "Collapse all" in other applications is to expand or collapse the entire tree, regardless of the selected item(s).

Having another entry, "Collapse/Expand Children (recursively)", would do the trick yeah, doesn't need to be in this PR. Concerning the naming, "Collapse Recursively' vs "Collapse All", i'm ok with either.
